### PR TITLE
Upgrade: check systemvm template before db changes

### PR DESCRIPTION
**Pull request by ustcweizhou**
### Description

Currently in upgrade cloudstack checks the systemvm template after db changes.
if user do not install systemvm template and uprgade cloudstack (for example install packages with newer version), upgrade will fail. in this case it will be very complicated to rollback the db if there is no db backup.

This pr checks the latest (only the last) systemvm template in upgrade path, and exits if required systemvm template is not registered.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [X] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by ustcweizhou** [comment]
@davidjumani @rhtyd

**Comment by rhtyd** [comment]
@ustcweizhou isn't the whole db upgrade path wrapped in a transaction? I suppose the whole transaction would fail without committing anything to the DB?

**Comment by weizhouapache** [comment]
> @ustcweizhou isn't the whole db upgrade path wrapped in a transaction? I suppose the whole transaction would fail without committing anything to the DB?

@rhtyd 
unfortunately no...
In each upgrade, the transaction will be started, committed and finally closed.
https://github.com/apache/cloudstack/blob/master/engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java#L240-L310
it is bad, but not the worst.

new system template is checked in performMigration. Before that, sql in script will be committed
(1) autocommit is true.
in https://github.com/apache/cloudstack/blob/master/framework/db/src/main/java/com/cloud/utils/db/ScriptRunner.java#L148-L150
(2) autocommit is false.
https://github.com/apache/cloudstack/blob/master/framework/db/src/main/java/com/cloud/utils/db/ScriptRunner.java#L181-L183

options are
(1) check systemvm template at the beginning of whole upgrade, or
(2) check systemvm template at each version upgrade.

**Review by RodrigoDLopez — APPROVED** [review]
code lgtm

File: engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java [hunk 1]
```diff
@@ -237,6 +238,32 @@ protected void upgrade(CloudStackVersion dbVersion, CloudStackVersion currentVer
...
+                    break;
+                } catch (CloudRuntimeException e) {
+                    String errorMessage = "Unable to upgrade the database";
+                    s_logger.error(errorMessage, e);
+                    throw new CloudRuntimeException(errorMessage, e);
+                } finally {
+                    txn.close();
+                }
+            }
+        }
```
Comment (DaanHoogland) [inline engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java:265]: can we put this in a separate method?

**Review by DaanHoogland — APPROVED** [review]
good functionality, good code.

**Comment by DaanHoogland** [comment]
test proposal:
1. don't upload templates,
2. upgrade,
3. start,
4. stop,
5. downgrade,
6. start,
7. upload templates,
8. stop,
9. start.
agree?

File: engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java (hunk 1 above)
Comment (weizhouapache) [reply engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java:261]: @DaanHoogland yes, done

**Comment by weizhouapache** [comment]
> test proposal:
> 
> 1. don't upload templates,
> 2. upgrade,
> 3. start,
> 4. stop,
> 5. downgrade,
> 6. start,
> 7. upload templates,
> 8. stop,
> 9. start.
>    agree?

@DaanHoogland yes, exactly

**Comment by weizhouapache** [comment]
@DaanHoogland @rhtyd 
can you kick off trillian test ?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
jenkins build failed at point below. close/open this pr to retry.

```
[INFO] --< org.apache.cloudstack:cloud-plugin-integrations-cloudian-connector >--
[INFO] Building Apache CloudStack Plugin - Cloudian Connector 4.15.1.0-SNAPSHOT [74/116]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-checkstyle-plugin:3.1.0:check (cloudstack-checkstyle) @ cloud-plugin-integrations-cloudian-connector ---
[INFO] Starting audit...
Audit done.
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.3:prepare-agent (prepare-coverage-agent) @ cloud-plugin-integrations-cloudian-connector ---
[INFO] argLine set to -javaagent:/home/jenkins/.m2/repository/org/jacoco/org.jacoco.agent/0.8.3/org.jacoco.agent-0.8.3-runtime.jar=destfile=/home/jenkins/jenkins-agent/workspace/Cloudstack/cloudstack-pr-analysis/plugins/integrations/cloudian/target/jacoco.exec
[INFO] 
[INFO] --- maven-remote-resources-plugin:1.3:process (default) @ cloud-plugin-integrations-cloudian-connector ---
[INFO] 
[INFO] --- maven-resources-plugin:2.5:resources (default-resources) @ cloud-plugin-integrations-cloudian-connector ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] Copying 3 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ cloud-plugin-integrations-cloudian-connector ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 10 source files to /home/jenkins/jenkins-agent/workspace/Cloudstack/cloudstack-pr-analysis/plugins/integrations/cloudian/target/classes
[INFO] 
[INFO] --- maven-resources-plugin:2.5:testResources (default-testResources) @ cloud-plugin-integrations-cloudian-connector ---
[debug] execute contextualize
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/jenkins/jenkins-agent/workspace/Cloudstack/cloudstack-pr-analysis/plugins/integrations/cloudian/src/test/resources
[INFO] Copying 3 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ cloud-plugin-integrations-cloudian-connector ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /home/jenkins/jenkins-agent/workspace/Cloudstack/cloudstack-pr-analysis/plugins/integrations/cloudian/target/test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.22.2:test (default-test) @ cloud-plugin-integrations-cloudian-connector ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.cloudstack.cloudian.CloudianClientTest
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
log4j:WARN No appenders could be found for logger (org.apache.cloudstack.cloudian.client.CloudianClient).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
[ERROR] Tests run: 25, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 9.1 s <<< FAILURE! - in org.apache.cloudstack.cloudian.CloudianClientTest
[ERROR] addUserAccountFail(org.apache.cloudstack.cloudian.CloudianClientTest)  Time elapsed: 5.447 s  <<< ERROR!
org.apache.cloudstack.api.ServerApiException: Operation timed out, please try again.
	at org.apache.cloudstack.cloudian.CloudianClientTest.addUserAccountFail(CloudianClientTest.java:145)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   CloudianClientTest.addUserAccountFail:145 » ServerApi Operation timed out, ple...
[INFO] 
[ERROR] Tests run: 25, Failures: 0, Errors: 1, Skipped: 0
```

File: engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java [hunk 2]
```diff
@@ -232,11 +233,41 @@ protected void runScript(Connection conn, InputStream file) {
...
+                        throw new CloudRuntimeException(errorMessage, e);
+                    }
+                    ((DbUpgradeSystemVmTemplate)upgrade).updateSystemVmTemplates(conn);
+                    break;
+                } catch (CloudRuntimeException e) {
+                    String errorMessage = "Unable to upgrade the database";
+                    s_logger.error(errorMessage, e);
+                    throw new CloudRuntimeException(errorMessage, e);
+                } finally {
+                    txn.close();
```
Comment (Pearl1594) [inline engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java:258]: @weizhouapache I believe there's a need to do a txn.commit() before closing the transaction as the DB changes made as part of the updateSystemVmTemplate() gets rolled back
```
Rolling back the transaction: Time = 22 Name =  Upgrade; called by -TransactionLegacy.rollback:888-TransactionLegacy.removeUpTo:831-TransactionLegacy.close:655-DatabaseUpgradeChecker.updateSystemVmTemplates:264-DatabaseUpgradeChecker.upgrade:275-DatabaseUpgradeChecker.check:379-CloudStackExtendedLifeCycle.checkIntegrity:64-CloudStackExtendedLifeCycle.start:54-DefaultLifecycleProcessor.doStart:182-DefaultLifecycleProcessor.access$200:53-DefaultLifecycleProcessor$LifecycleGroup.start:360-DefaultLifecycleProcessor.startBeans:158
2021-02-01 09:50:57,795 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41500to41510 to upgrade from 4.15.0.0-4.15.1.0 to 4.15.1.0
```

Comment (weizhouapache) [reply engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java:258]: > @ustcweizhou I believe there's a need to do a txn.commit() before closing the transaction as the DB changes made as part of the updateSystemVmTemplate() gets rolled back
> 
> ```
> Rolling back the transaction: Time = 22 Name =  Upgrade; called by -TransactionLegacy.rollback:888-TransactionLegacy.removeUpTo:831-TransactionLegacy.close:655-DatabaseUpgradeChecker.updateSystemVmTemplates:264-DatabaseUpgradeChecker.upgrade:275-DatabaseUpgradeChecker.check:379-CloudStackExtendedLifeCycle.checkIntegrity:64-CloudStackExtendedLifeCycle.start:54-DefaultLifecycleProcessor.doStart:182-DefaultLifecycleProcessor.access$200:53-DefaultLifecycleProcessor$LifecycleGroup.start:360-DefaultLifecycleProcessor.startBeans:158
> 2021-02-01 09:50:57,795 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41500to41510 to upgrade from 4.15.0.0-4.15.1.0 to 4.15.1.0
> ```

@Pearl1594 thanks for testing.
could you tell how to reproduce the issue ?

Comment (Pearl1594) [reply engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java:258]: @weizhouapache I tested this by merging this code to master. However, the steps are equivalent to :
1. Deploy a 4.14 env
2. Register the template (4.15 template)
3. upgrade to 4.15
4. The Management server comes up fine, but, if we check the cloud.configuration value (in the DB) for attribute - "router.template.<hypervisor>" notice it doesn't update the template name to the new one. 

The case of not applying the DB changes when template isn't registered, works absolutely fine.

Comment (weizhouapache) [reply engine/schema/src/main/java/com/cloud/upgrade/DatabaseUpgradeChecker.java:259]: @Pearl1594 yeah, you are right !
pushed  a change. 
thanks for review and testing.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Review by Pearl1594 — APPROVED** [review]
LGTM

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
I think we don't need this as the upgrade was tested manually by @Pearl1594 . better safe though
@blueorangutan test

**Review by shwstppr — APPROVED** [review]
LGTM

**Review by rhtyd — APPROVED** [review]

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41400to41500.java [hunk 3]
```diff
@@ -67,12 +67,12 @@ public boolean supportsRollingUpgrade() {
 
     @Override
     public void performDataMigration(Connection conn) {
-        updateSystemVmTemplates(conn);
```
Comment (rhtyd) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41400to41500.java:70]: +1 important thing to remember now is not to run this again

**Pull request merged**